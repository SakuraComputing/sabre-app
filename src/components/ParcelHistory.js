import React from 'react';
import parcelData from '../../data/parcel.json';
import trackData from '../../data/tracks.json';
import addressData from '../../data/address.json';
import ParcelDetails from "./ParcelDetails";
import TrackingHistory from "./TrackingHistory";

class ParcelHistory extends React.Component {

    constructor(props) {
        super(props);
        this.state = {parcelId: '', error: '', selectedParcel: '', selectedTracks: '', selectedAddress: ''};
    }

    onSubmit = (e) => {
        e.preventDefault();

        let parcelId = this.state.parcelId;
        if (parcelId) {
            if(this.onValidate()) {
                const selectedParcel = parcelData["parcels"].find(parcel => parcel.parcelId === parcelId);
                if (selectedParcel) {
                    let addressNo = selectedParcel.addressNo;
                    var selectedAddress = addressData["addresses"].find(address => address.addressNo === addressNo);
                    var selectedTracks = trackData["tracks"].filter(track => track.parcelId === parcelId);
                }
                let error = selectedParcel ? null : 'parcel not found';

                this.setState(() => ({
                    selectedParcel: selectedParcel,
                    error: error,
                    selectedTracks: selectedTracks,
                    selectedAddress: selectedAddress
                }));
            }
        } else {
            this.setState(() => ({ error: 'Please provide a parcel Id' }));
        }
    };

    onParcelChange = (e) => {
        const parcelId = e.target.value;
        this.setState(() => ({ parcelId }));
    };

    onClear = () => {
        this.setState(()=> ({ parcelId: '', error: null, selectedParcel: '', selectedTracks: '', selectedAddress: '' }))
    };

    onValidate = (e) => {
        let parcelId = this.state.parcelId;
        let formIsValid = true;

        if(parcelId.length < 12 || parcelId.length > 35) {
            this.setState(() => ({ error: 'Parcel must be between 12 and 35 chars' }));
            formIsValid = false
        }
        return formIsValid;
    }
    ;

    render() {
        let id = 'parcelId';
        return (
            <div>
                <form className="quick-form" onSubmit={this.onSubmit} >
                    <h2>Parcel History</h2>
                    <div className="search-bar">
                        Parcel Id:
                        <input
                            onChange={this.onParcelChange}
                            name={id}
                            id={id}
                            value={this.state.parcelId}
                            placeholder="Parcel Id"
                        />
                        <button className="submit-button">
                            Search
                        </button>
                        <button type="reset" className="clear-button" onClick={this.onClear}>
                            Clear Search
                        </button>
                    </div>
                    {this.state.selectedParcel ? <ParcelDetails parcel={this.state.selectedParcel} address={this.state.selectedAddress}/> : ''}
                    {this.state.selectedTracks ? <TrackingHistory tracks={this.state.selectedTracks} /> : ''}
                </form>
                {this.state.error && <div className="error-message">{this.state.error}</div>}
            </div>
        )
    };
};
export default ParcelHistory;

