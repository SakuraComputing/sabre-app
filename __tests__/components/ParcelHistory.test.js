import React from 'react';
import {configure, shallow} from 'enzyme';
import ParcelHistory from '../../src/components/ParcelHistory';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('ParcelHistoryTests', () => {

    let testComponent;

    beforeEach(() => {
        testComponent = shallow(<ParcelHistory/>);
    });

    it('should render Parcel History correctly', () => {
        expect(testComponent).toMatchSnapshot();
    });

    it('should return an error when no parcel number is entered', () => {
        testComponent.find('form').simulate('submit', {
            preventDefault: () => {
            }
        });
        expect(testComponent.state('error').length).toBeGreaterThan(0);
        expect(testComponent).toMatchSnapshot();
    });

    it('should set the state to all clear',() => {
        // testComponent.find('form').simulate('clear');
        let mockPreventDefault = jest.fn();
        let mockEvent = {preventDefault: mockPreventDefault};
        testComponent.instance().onClear(mockEvent);
        expect(testComponent.state()).toEqual({"error": null, "parcelId": "", "selectedParcel": "", "selectedTracks": "", "selectedAddress": ''});
    });

    it('should set the state onChange to parcel id', () => {
        let mockEvent = {target: {name: 'parcelId', value: null}};
        let parcelId = 'ParcelID';
        mockEvent.target['value'] = parcelId;
        testComponent.instance().onParcelChange(mockEvent);
        expect(testComponent.state().parcelId).toEqual(parcelId);
    });

    describe('onSubmit is executed', () => {
        let mockPreventDefault = jest.fn();
        let mockEvent = {preventDefault: mockPreventDefault};

        it('should prevent the default event from firing', () => {
            testComponent.instance().onSubmit(mockEvent);
            expect(mockPreventDefault).toHaveBeenCalled();
        });
        it('should display an error when no parcel selected',() => {
            testComponent.setState({ parcelId: ''});
            testComponent.instance().onSubmit(mockEvent);
            expect(testComponent.state().error).toBe('Please provide a parcel Id');
        });
        it('should display as error when parcel is not found',() => {
           testComponent.setState({ parcelId: '123456789012'});
            testComponent.instance().onSubmit(mockEvent);
            expect(testComponent.state().error).toBe('parcel not found');
        });
        it('should find tracks for a valid parcel',() => {
            testComponent.setState({ parcelId: 'JD0002256504047335'});
            testComponent.instance().onSubmit(mockEvent);
            expect(testComponent.state('selectedTracks').length).toBeGreaterThan(0);
        });
        it('should validate the parcel field',() => {
            testComponent.setState({parcelId: 'wibble'});
            testComponent.instance().onSubmit(mockEvent);
            expect(testComponent.state().error).toBe('Parcel must be between 12 and 35 chars');
        });
    });
});


