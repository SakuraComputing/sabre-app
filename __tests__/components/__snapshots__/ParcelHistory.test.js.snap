// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParcelHistoryTests should render Parcel History correctly 1`] = `
<div>
  <form
    className="quick-form"
    onSubmit={[Function]}
  >
    <h2>
      Parcel History
    </h2>
    <div
      className="search-bar"
    >
      Parcel Id:
      <input
        id="parcelId"
        name="parcelId"
        onChange={[Function]}
        placeholder="Parcel Id"
        value=""
      />
      <button
        className="submit-button"
      >
        Search
      </button>
      <button
        className="clear-button"
        onClick={[Function]}
        type="reset"
      >
        Clear Search
      </button>
    </div>
  </form>
</div>
`;

exports[`ParcelHistoryTests should return an error when no parcel number is entered 1`] = `
<div>
  <form
    className="quick-form"
    onSubmit={[Function]}
  >
    <h2>
      Parcel History
    </h2>
    <div
      className="search-bar"
    >
      Parcel Id:
      <input
        id="parcelId"
        name="parcelId"
        onChange={[Function]}
        placeholder="Parcel Id"
        value=""
      />
      <button
        className="submit-button"
      >
        Search
      </button>
      <button
        className="clear-button"
        onClick={[Function]}
        type="reset"
      >
        Clear Search
      </button>
    </div>
  </form>
  <div
    className="error-message"
  >
    Please provide a parcel Id
  </div>
</div>
`;
